### PR TITLE
Add GCHeapMallocator

### PR DESCRIPTION
**Pull request by pbackus**
In preparation for the removal of GCAllocator.deallocate and
GCAllocator.reallocate, make that functionality available via a new
allocator with a different name.

Users of GCAllocator that wish to continue using manual `@system` memory
management instead of automatic `@safe` memory management can switch to
GCHeapMallocator.

For the rationale behind these changes, see PR #8554 and issue 23318.

---

CC @CyberShadow @schveiguy 

@atilaneves this adds a new symbol

### CODE REVIEW

**Comment by pbackus** [comment]
```
core.exception.AssertError@std/logger/core.d(1905): foo
```
Looks like a flaky test. Force pushing to retry...

**Comment by burner** [comment]
https://github.com/dlang/phobos/pull/8557

**Comment by pbackus** [comment]
Rebased to pick up #8557

**Comment by schveiguy** [comment]
Can this be an alias/wrapper underneath? Hard to review this when most likely it's simply a copy-paste of existing code.

**Comment by pbackus** [comment]
@schveiguy Made it a wrapper.

File: std/experimental/allocator/gc_allocator.d [hunk 1]
```diff
@@ -197,3 +197,146 @@ pure nothrow @safe unittest
...
+    /**
+    Standard allocator methods per the semantics defined above. The $(D
+    deallocate) and `reallocate` methods are `@system` because they may
+    move memory around, leaving dangling pointers in user code.
+    */
+    pure nothrow @trusted void[] allocate(size_t bytes) shared const
+    {
+        return impl.allocate(bytes);
+    }
+
```
Comment (CyberShadow) [inline std/experimental/allocator/gc_allocator.d:225]: Would `alias this` work here, instead of explicit forwards?

Comment (pbackus) [reply std/experimental/allocator/gc_allocator.d:225]: `alias this` to a `private` field doesn't work when used from outside the module, and even if it did, I don't think the implicit conversion would be desirable.

File: std/experimental/allocator/gc_allocator.d [hunk 2]
```diff
@@ -197,3 +197,194 @@ pure nothrow @safe unittest
     void *badPtr = (() @trusted => cast(void*)(0xdeadbeef))();
     assert(GCAllocator.instance.resolveInternalPointer(badPtr, result) == Ternary.no);
 }
+
+/**
+Manual allocator that allocates from the D runtime's garbage-collected heap.
```
[the remark this answers is not in the document]
Comment (pbackus) [reply std/experimental/allocator/gc_allocator.d]: Updated the docs for `GCHeapMallocator` here; will do the same for `GCAllocator` in the other PR.

**Comment by atilaneves** [comment]
Please see my comments on the other PR.

**Pull request closed without merging**